server/src/main/java/keywhiz/Tracing.java

```diff
@@ -0,0 +1,91 @@
+package keywhiz;
+
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import io.opentracing.log.Fields;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.GlobalTracer;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public final class Tracing {
+  private Tracing() {}
+
+  public static <T> T trace(String spanName, Supplier<T> delegate) {
+    Tracer tracer = GlobalTracer.get();
+
+    Span span = tracer.buildSpan(spanName).start();
+    try {
+      try (Scope ignored = tracer.scopeManager().activate(span)) {
+        return delegate.get();
+      } catch (Exception e) {
+        tagErrors(span, e);
+        throw e;
+      }
+    } finally {
+      span.finish();
+    }
+  }
+
+  public static <T> T tagErrors(Supplier<T> delegate) {
+    Span span = GlobalTracer.get().activeSpan();
+    try {
+      return delegate.get();
+    } catch (Exception e) {
+      tagErrors(span, e);
+      throw e;
+    }
+  }
+
+  private static void tagErrors(Span span, Exception e) {
+    span.setTag(Tags.ERROR, true);
+
+    tagErrorsOpenTracing(span, e);
+    tagErrorsDataDog(span, e);
+  }
+
+  private static void tagErrorsOpenTracing(Span span, Exception e) {
+    https://github.com/opentracing/specification/blob/11dd7f8b16739f02761a9541f04e39f27738f525/semantic_conventions.md#span-and-log-errors
+
+    span.log(
```

Comment:
The fact that URLs are legal Java syntax always cracks me up

Reply:
No way! TIL

Reply:
It's because the protocol is interpreted as a label, which you tend not to see too much of, and everything after the slashes is interpreted as a comment. Of course, since labels need to be unique within a method, you can't just go crazy, and this is definitely an abuse of syntax, but still fun trivia.